src/logger/instrumentation.ts

```diff
@@ -0,0 +1,95 @@
+import { logs } from '@opentelemetry/api-logs';
+import { LoggerProvider, BatchLogRecordProcessor } from '@opentelemetry/sdk-logs';
+import { defaultResource } from '@opentelemetry/resources';
+import { logExporter } from './circular-buffer-exporter';
+
+let isLoggingInitialized = false;
+
+type ConsoleMethods = 'log' | 'info' | 'warn' | 'error' | 'debug';
+
+export function initializeLogging(): void {
+  const isBrowser =
+    typeof globalThis !== 'undefined' &&
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    typeof (globalThis as any).window !== 'undefined';
+
+  if (isLoggingInitialized || isBrowser) {
+    return;
+  }
+
+  try {
+    const resource = defaultResource();
+
+    const loggerProvider = new LoggerProvider({
+      resource,
+      processors: [
+        new BatchLogRecordProcessor(logExporter, {
+          maxQueueSize: 100,
+          maxExportBatchSize: 10,
+          scheduledDelayMillis: 1000,
+        }),
+      ],
+    });
+
+    logs.setGlobalLoggerProvider(loggerProvider);
+
+    // Patch console methods to emit OTel logs
+    patchConsole();
+
+    isLoggingInitialized = true;
+  } catch (error) {
+    console.error('[Tidewave] Failed to initialize logging:', error);
+  }
+}
+
+function patchConsole(): void {
+  const logger = logs.getLogger('console', '1.0.0');
+
+  const severityMap: Record<ConsoleMethods, string> = {
+    log: 'INFO',
+    info: 'INFO',
+    warn: 'WARN',
+    error: 'ERROR',
+    debug: 'DEBUG',
+  };
+
+  (Object.entries(severityMap) as [ConsoleMethods, string][]).forEach(
+    ([method, severity]): void => {
+      const original = console[method].bind(console);
+
+      console[method] = (...args: unknown[]): void => {
+        original(...args);
+
+        try {
+          const body = args
+            .map((arg: unknown) => {
+              if (typeof arg === 'string') return arg;
+              if (arg instanceof Error) {
+                return `${arg.name}: ${arg.message}\n${arg.stack}`;
```

Comment:
[`arg.stack`](https://nodejs.org/api/errors.html#errorstack) already formats the name and message:

```suggestion
                return arg.stack;
```

In browser environment it's not exactly stanrdized, but in Node.js it should be more reliable.